crypto-common: add SerializableState trait

rpiasetskyi opened the pull request:
This trait is used for saving internal state of the hash core and
restoring it later.

The original issue was discussed here: https://github.com/RustCrypto/hashes/issues/310

tarcieri commented on digest/src/core_api.rs:
```diff
@@ -117,3 +117,15 @@ pub enum TruncSide {
 }
+
+/// Core trait for saving internal state of the hash core and restoring it later.
+pub trait HashCoreInternalState {
```
I'd suggest using Rust's built-in conversion traits rather than defining special methods like `internal_state` and `from_internal_state`:

```suggestion
pub trait HashCoreInternalState: TryFrom<Self::InternalState, Error = Error> + Into<Self::InternalState> {
```

tarcieri commented:
I think the core notion here is a "serializable state" rather than an "internal state"

tarcieri commented on digest/src/core_api.rs:
```diff
@@ -117,3 +117,15 @@ pub enum TruncSide {
+/// Core trait for saving internal state of the hash core and restoring it later.
+pub trait HashCoreInternalState {
+    /// Internal state of the hash core.
+    type InternalState;
```
It'd be good to have some bounds here to enable generic code:

```suggestion
    type InternalState: AsRef<[u8]>;
```

rpiasetskyi replied:
If the internal state consists of several objects, is it possible to implement `TryFrom<&[u8]> + AsRef<[u8]>`: e.g. internal state of SHA-2 hashes consists of the intermediate hash value (array of `u32`/`u64`) and the number of consumed blocks (number `u64`/`u128`).

tarcieri replied:
The use case of this API is serialization.

If you think there's a use case beyond serialization, can you explain how that use case isn't satisfied by simply cloning an existing digest instance?

rpiasetskyi replied:
Yes, the use case is serialization only.
Should I then create a byte array and serialize the internal state on my own?
E.g. for `Sha256VarCore`, I need to allocate an array, copy the intermediate hash value (`[u32; STATE_LEN]`), and the number of consumed blocks (`u64`).
Or is it better to return an object that consists of the intermediate hash value and the number of consumed blocks, and this object is `Serializable` (`serde`)? But this object can not implement `AsRef<[u8]>` because it consists of a byte array and number.

tarcieri replied:
It depends if the state can always be fixed-width or not. If it can, we should probably use an associated `generic_array::ArrayLength<u8>` instead.

I think it's a pretty reasonable assumption that most digests could support a fixed-width serializable state.

newpavlov reviewed (commented):
I agree that it's worth to introduce trait which outputs generic array. Something like this:
```rust
pub type SerializedState<T> = GenericArray<u8, <T as SerializableState>::SerializedStateSize>;

pub trait SerializableState {
    type SerializedStateSize: ArrayLength<u8>;
    
    fn serialize(&self) -> SerializedState<Self>;
    fn deserialize(buf: &SerializedState<Self>) -> Self;
}
```
Also do not forget to add warning that serialized state may contain sensitive information, e.g. in the hashes case it can contain parts of unprocessed message.

The trait may go to the `crypto-common` crate since it may be useful for other algorithms as well.

rpiasetskyi commented:
Sorry, missed the last comment. I will move the trait to `crypto-common` and use `serialize/deserialize` methods.

Should `deserialize` method return `Result` (if the buffer contains invalid values)?

rpiasetskyi commented:
The versions of the crates have to be updated: should they be `0.1.7/0.10.4` or `0.2.0/0.11.0`?

Right now crypto-common is `0.1.6` and digest is `0.10.3`.

newpavlov commented on digest/src/core_api/ct_variable.rs:
```diff
@@ -165,3 +165,26 @@ where
     }
 }
+
+impl<T, OutSize> SerializableState for CtVariableCoreWrapper<T, OutSize>
```
Do not forget to also implement it for `CoreWrapper` and ideally for `RtVariableCoreWrapper`.

rpiasetskyi replied:
`CoreWrapper` has [`decompose`](https://docs.rs/digest/0.10.3/digest/core_api/struct.CoreWrapper.html#method.decompose) method that returns buffered input and `CtVariableCoreWrapper` object, and [`from_core`](https://docs.rs/digest/0.10.3/digest/core_api/struct.CoreWrapper.html#method.from_core) method to restore the object.

Do we need additionally implement `SerializableState`? Also, we do not know the size of buffered input (or we need to store the length with the partially filled block of buffered input).

newpavlov replied:
I think, yes. It would be weird to support `SerializableState` only for the core types and force users to jump through artificial error-prone hoops.

> Also, we do not know the size of buffered input (or we need to store the length with the partially filled block of buffered input).

You should attach to serialized state number of unprocessed bytes and zero-padded unprocessed data (you could use `buffer.clone().pad_with_zeros().clone()`). For SHA-256 it would mean always attaching 65 bytes, even when there are no unprocessed bytes in it. We could save one byte for the eager hashes by using the fact that the last byte will be never used, but I think it's too small of a win for the very niche feature, so it does not worth to introduce the additional complexity.

newpavlov reviewed (commented):
>Should deserialize method return Result (if the buffer contains invalid values)?

Yes, I forgot to do it in my snippet. I think we can define the error type in `crypto-common` (i.e. no need to make it generic).

>The versions of the crates have to be updated: should they be 0.1.7/0.10.4 or 0.2.0/0.11.0?

The changes are backwards-compatible, so patch releases should be fine.

newpavlov commented on crypto-common/src/lib.rs:
```diff
@@ -309,3 +309,34 @@ impl fmt::Display for InvalidLength {
+#[cfg(feature = "std")]
+impl std::error::Error for DeserializeStateError {}
+
+/// Types which can serialize the internal state and be restored from it.
```
Do not forget to add the warning that serialized state may contain sensitive data in plain text.

tarcieri commented:
The serialized state needs to be consistent regardless of the target pointer width.

You should pick a sensible on-the-wire representation (`u32` or `u64`) and use a `try_from` conversion when deserializing it.

newpavlov commented:
I think we can simply transform `output_size` to `u8` using `TryInto` and `unwrap`. I am not aware about variable hashes which support output size bigger than 255 bytes (XOF does not count). In future versions it could be worth to replace `usize` with `u8`.

newpavlov commented on digest/src/core_api/ct_variable.rs:
```diff
@@ -165,3 +166,25 @@ where
+
+    fn serialize(&self) -> SerializedState<Self> {
+        self.inner.serialize()
+    }
```
As a sanity check it could be worth to add into serialized state output size (the `OutSize` type parameter) and return an error during deserialization if it does not match.

newpavlov commented on digest/src/core_api/rt_variable.rs:
```diff
@@ -145,6 +149,68 @@ where
+    <T::SerializedStateSize as Add<U9>>::Output: Add<T::BlockSize>,
+    <<T::SerializedStateSize as Add<U9>>::Output as Add<T::BlockSize>>::Output: ArrayLength<u8>,
+{
+    type SerializedStateSize = Sum<Sum<T::SerializedStateSize, U9>, T::BlockSize>;
```
I think you forgot to change `U9` to `U2`.

newpavlov commented on digest/src/core_api/rt_variable.rs:
```diff
@@ -145,6 +149,68 @@ where
+
+        Ok(Self {
+            core: T::deserialize(serialized_core)?,
+            buffer: BlockBuffer::new(&block_buffer[..pos]),
```
Note that if the buffer is "eager" and `pos` is equal to block size, `new` will panic. I guess the simplest solution will be to add `try_new` method to `BlockBuffer`.

rpiasetskyi replied:
https://github.com/RustCrypto/utils/pull/799

newpavlov commented on digest/src/core_api/wrapper.rs:
```diff
@@ -227,6 +228,61 @@ where
+
+        Ok(Self {
+            core: T::deserialize(serialized_core)?,
+            buffer: BlockBuffer::new(block_buffer),
```
Same as in the previous comment.

newpavlov reviewed (commented):
You could use the [`concat`](https://docs.rs/generic-array/latest/generic_array/struct.GenericArray.html#method.concat) method to slightly improve the code, though it could force you to make `SerializedStateSize` less... pretty. So it's quite optional.

rpiasetskyi commented:
Converted to draft to avoid emails about failed tests.

newpavlov commented on digest/Cargo.toml:
```diff
@@ -15,7 +15,7 @@ categories = ["cryptography", "no-std"]
 
 # optional dependencies
-block-buffer = { version = "0.10", optional = true }
+block-buffer = { version = ">=0.10.3", optional = true }
```
I think `0.10.3` should be sufficient, i.e. the `>=` part is redundant.

newpavlov commented on digest/src/core_api/rt_variable.rs:
```diff
@@ -145,6 +156,54 @@ where
+    <<<<<T::SerializedStateSize as Add<U1>>::Output as Add<T::BlockSize>>::Output as Add<U1>>::Output as Sub<T::SerializedStateSize>>::Output as Sub<U1>>::Output: Sub<T::BlockSize> + ArrayLength<u8>,
+    <<<<<<T::SerializedStateSize as Add<U1>>::Output as Add<T::BlockSize>>::Output as Add<U1>>::Output as Sub<T::SerializedStateSize>>::Output as Sub<U1>>::Output as Sub<T::BlockSize>>::Output: ArrayLength<u8>,
+{
+    type SerializedStateSize = Sum<Sum<Sum<T::SerializedStateSize, U1>, T::BlockSize>, U1>;
```
You can add ignore attribute for this associated type to work around the clippy warnings.

rpiasetskyi replied:
I have checked that the current version of `clippy` doesn't treat this as a warning anymore. The first version that does it is `1.59.0`.
Does it make sense to update the version of clippy (to the current stable `1.62.0` or minimal version `1.59.0`)? Or is it better to add an attribute?

newpavlov replied:
Yes, you can update clippy's version in CI.

newpavlov commented on digest/src/core_api/rt_variable.rs:
```diff
@@ -145,6 +156,54 @@ where
+    <<<T::SerializedStateSize as Add<U1>>::Output as Add<T::BlockSize>>::Output as Add<U1>>::Output: ArrayLength<u8> + Sub<T::SerializedStateSize>,
+    <<<<T::SerializedStateSize as Add<U1>>::Output as Add<T::BlockSize>>::Output as Add<U1>>::Output as Sub<T::SerializedStateSize>>::Output: Sub<U1> + ArrayLength<u8>,
+    <<<<<T::SerializedStateSize as Add<U1>>::Output as Add<T::BlockSize>>::Output as Add<U1>>::Output as Sub<T::SerializedStateSize>>::Output as Sub<U1>>::Output: Sub<T::BlockSize> + ArrayLength<u8>,
+    <<<<<<T::SerializedStateSize as Add<U1>>::Output as Add<T::BlockSize>>::Output as Add<U1>>::Output as Sub<T::SerializedStateSize>>::Output as Sub<U1>>::Output as Sub<T::BlockSize>>::Output: ArrayLength<u8>,
```
I expected something similar, but still... :)

I think you can simplify it a fair bit by using [operator aliases](https://docs.rs/typenum/latest/typenum/operator_aliases/index.html) in bounds as well, e.g. you can replace:
```rust
<T::SerializedStateSize as Add<U1>>::Output: Add<T::BlockSize> + ArrayLength<u8>,
```
with:
```rust
Sum<T::SerializedStateSize, U1>: Add<T::BlockSize> + ArrayLength<u8>,
```

rpiasetskyi commented:
Does it make sense to implement `SerializableState` for some commonly used types?

```
impl SerializableState for u8 {
    type SerializedStateSize = U1;

    fn serialize(&self) -> SerializedState<Self> {
        [*self].into()
    }

    fn deserialize(
        serialized_state: &SerializedState<Self>,
    ) -> Result<Self, DeserializeStateError> {
        Ok(serialized_state[0])
    }
}

impl SerializableState for u64 {
    type SerializedStateSize = U8;

    fn serialize(&self) -> SerializedState<Self> {
        self.to_le_bytes().into()
    }

    fn deserialize(
        serialized_state: &SerializedState<Self>,
    ) -> Result<Self, DeserializeStateError> {
        Ok(u64::from_le_bytes((*serialized_state).into()))
    }
}

impl SerializableState for [u64; 16] {
    type SerializedStateSize = U128;

    fn serialize(&self) -> SerializedState<Self> {
        let mut serialized_state = SerializedState::<Self>::default();
        for (val, chunk) in self.iter().zip(serialized_state.chunks_exact_mut(8)) {
            chunk.copy_from_slice(&val.to_le_bytes());
        }

        serialized_state
    }

    fn deserialize(
        serialized_state: &SerializedState<Self>,
    ) -> Result<Self, DeserializeStateError> {
        let mut array = Self::default();

        for (val, chunk) in array.iter_mut().zip(serialized_state.chunks_exact(8)) {
            *val = u64::from_le_bytes(chunk.try_into().unwrap());
        }

        Ok(array)
    }
}
```

It can be implemented in a similar way to this https://github.com/fizyk20/generic-array/blob/master/src/impls.rs#L183.

This can move the for loop (array serialization/deserialization) in one place and make the code a little bit simpler.

```
type State = [u64; 16];

impl SerializableState for GroestlLongVarCore {
    type SerializedStateSize = Sum<
        <u64 as SerializableState>::SerializedStateSize,
        <State as SerializableState>::SerializedStateSize,
    >;

    fn serialize(&self) -> SerializedState<Self> {
        self.state.serialize().concat(self.blocks_len.serialize())
    }

    fn deserialize(
        serialized_state: &SerializedState<Self>,
    ) -> Result<Self, DeserializeStateError> {
        let (serialized_state, serialized_blocks_len) =
            Split::<_, <State as SerializableState>::SerializedStateSize>::split(serialized_state);

        Ok(Self {
            state: State::deserialize(serialized_state)?,
            blocks_len: u64::deserialize(serialized_blocks_len)?,
        })
    }
}
```

rpiasetskyi commented:
With the introduction of the `AssociatedOid` trait and new generic parameter for `CtVariableCoreWrapper`, is it needed to add it to serialized state?

tarcieri commented:
No, those are type-level properties which are identical regardless of the state.

rpiasetskyi commented:
> Does it make sense to implement `SerializableState` for some commonly used types?
> 
> ```
> impl SerializableState for u8 {
>     type SerializedStateSize = U1;
> 
>     fn serialize(&self) -> SerializedState<Self> {
>         [*self].into()
>     }
> 
>     fn deserialize(
>         serialized_state: &SerializedState<Self>,
>     ) -> Result<Self, DeserializeStateError> {
>         Ok(serialized_state[0])
>     }
> }
> 
> impl SerializableState for u64 {
>     type SerializedStateSize = U8;
> 
>     fn serialize(&self) -> SerializedState<Self> {
>         self.to_le_bytes().into()
>     }
> 
>     fn deserialize(
>         serialized_state: &SerializedState<Self>,
>     ) -> Result<Self, DeserializeStateError> {
>         Ok(u64::from_le_bytes((*serialized_state).into()))
>     }
> }
> 
> impl SerializableState for [u64; 16] {
>     type SerializedStateSize = U128;
> 
>     fn serialize(&self) -> SerializedState<Self> {
>         let mut serialized_state = SerializedState::<Self>::default();
>         for (val, chunk) in self.iter().zip(serialized_state.chunks_exact_mut(8)) {
>             chunk.copy_from_slice(&val.to_le_bytes());
>         }
> 
>         serialized_state
>     }
> 
>     fn deserialize(
>         serialized_state: &SerializedState<Self>,
>     ) -> Result<Self, DeserializeStateError> {
>         let mut array = Self::default();
> 
>         for (val, chunk) in array.iter_mut().zip(serialized_state.chunks_exact(8)) {
>             *val = u64::from_le_bytes(chunk.try_into().unwrap());
>         }
> 
>         Ok(array)
>     }
> }
> ```
> 
> It can be implemented in a similar way to this https://github.com/fizyk20/generic-array/blob/master/src/impls.rs#L183.
> 
> This can move the for loop (array serialization/deserialization) in one place and make the code a little bit simpler.
> 
> ```
> type State = [u64; 16];
> 
> impl SerializableState for GroestlLongVarCore {
>     type SerializedStateSize = Sum<
>         <u64 as SerializableState>::SerializedStateSize,
>         <State as SerializableState>::SerializedStateSize,
>     >;
> 
>     fn serialize(&self) -> SerializedState<Self> {
>         self.state.serialize().concat(self.blocks_len.serialize())
>     }
> 
>     fn deserialize(
>         serialized_state: &SerializedState<Self>,
>     ) -> Result<Self, DeserializeStateError> {
>         let (serialized_state, serialized_blocks_len) =
>             Split::<_, <State as SerializableState>::SerializedStateSize>::split(serialized_state);
> 
>         Ok(Self {
>             state: State::deserialize(serialized_state)?,
>             blocks_len: u64::deserialize(serialized_blocks_len)?,
>         })
>     }
> }
> ```
> 
> And it might be possible to implement `#[derive(SerializableState)]` for types that do not need to check anything.

What do you think about https://github.com/RustCrypto/traits/pull/1078/commits/3447449814c3ee2bae77834bd23d1984895603fa?
If this is okay, I will also add `#[derive(SerializableState)]`.

waynr commented:
:wave: any chance of getting this merged any time in the near future?

tarcieri commented:
@waynr it needs a rebase, at the very least

waynr commented:
@tarcieri I'm attempting to rebase but running into compile errors that are pretty much totally incomprehensible to me. This repo is like a firehose of generic types and my puny little brain only has capacity for a trickle.

tarcieri commented:
I think @rpiasetskyi said he would take a look

waynr commented:
Just a quick note - it looks like the issue of the local dependency on crypto-common being broken in master branch for the digest crate is fixed in #1358 so this PR might need to wait for that before being rebased.

rpiasetskyi commented:
I am doing a rebase (and migration to `hybrid_array`) using https://github.com/RustCrypto/traits/pull/1358.

rpiasetskyi commented:
@tarcieri Sorry, I was expecting it to update the PR but not to merge it :(

rpiasetskyi commented:
Reopen.

tarcieri commented:
I think you pushed 0 commits which auto-closed the PR? Regardless I can't reopen it... I think you'd need to push some new commits
